src/Streamly/Internal/Data/Ring/Foreign.hs

```diff
@@ -506,3 +508,48 @@ unsafeFoldRingNM count rh f z rb@Ring {..} =
         if ptr == rh || n == 0
             then return acc'
             else go (n - 1) acc' ptr
+
+data Tuple4' a b c d = Tuple4' !a !b !c !d deriving Show
+
+-- | @rollWindow n collector@ is a simple, incremental sliding window 
+-- fold that does not require all the intermediate elements in a computation.
+-- This maintains n elements in the window, when a new element comes it slides
+-- out the oldest element and the new element along with the old element are 
+-- supplied to the collector fold. 
+--
+-- The Maybe is for the case when initially the window is filling and 
+-- there is no old element.
+--
+{-# INLINE rollWindow #-}
+rollWindow :: forall m a b. (MonadIO m, Storable a) 
+    => Int -> Fold m (a, Maybe a) b -> Fold m a b
+rollWindow n (Fold step1 initial1 extract1)= Fold step initial extract
```

Comment:
We can name it as `slidingWindow`, that's a more common term for such processing, also it is a noun and folds are usually nouns.